srv/src/main/java/my/bookshop/handlers/CatalogServiceHandler.java

```diff
@@ -2,56 +2,223 @@
+	private final Messages messages;
+
+	private final BookRatingService bookRatingService;
+
+	private final CqnAnalyzer analyzer;
+	
+
+	CatalogServiceHandler(PersistenceService db, Messages messages, BookRatingService bookRatingService, CdsModel model) {
 		this.db = db;
+		this.messages = messages;
+		this.bookRatingService = bookRatingService;
+		this.analyzer = CqnAnalyzer.create(model);
+	}
+
+	/**
+	 * Invokes some validations before creating a review.
+	 * 
+	 * @param context {@link ReviewContext}
+	 */
+	@Before(entity = Books_.CDS_NAME)
+	public void beforeAddReview(AddReviewContext context) {
+		String reviewer = context.getUserInfo().getName();
+		String bookId = (String) analyzer.analyze(context.getCqn()).targetKeys().get(Books.ID);
+
+		Result result = db.run(Select.from(CatalogService_.REVIEWS)
+				.where(review -> review.book_ID().eq(bookId).and(review.reviewer().eq(reviewer))));
+
+		if (result.first().isPresent()) {
+			throw new ServiceException(ErrorStatuses.METHOD_NOT_ALLOWED,
+					"User not allowed to add more than one review for a given book").messageTarget(Reviews_.class,
+							r -> r.reviewer());
+		}
+	}
+
+	/**
+	 * Handles the review creation from the given context.
+	 * 
+	 * @param context {@link ReviewContext}
+	 */
+	@On(entity = Books_.CDS_NAME)
+	public void onAddReview(AddReviewContext context) {
+		Integer rating = context.getRating();
+		String title = context.getTitle();
+		String text = context.getText();
+
+		String username = context.getUserInfo().getName();
+		String bookId = (String) analyzer.analyze(context.getCqn()).targetKeys().get(Books.ID);
+
+		Reviews review = Reviews.create();
+		review.setBookId(bookId);
+		review.setReviewer(username);
+		review.setRating(rating);
+		review.setTitle(title);
+		review.setText(text);
+
+		validateReview(review);
+
+		Result res = db.run(Insert.into(CatalogService_.REVIEWS).entry(review));
+		Optional<Reviews> inserted = res.first(Reviews.class);
+
+		messages.success(MessageKeys.REVIEW_ADDED);
+		inserted.ifPresent(r -> context.setResult(r));
+	}
+
+	/**
+	 * Recalculates and sets the book rating after a new review for the given book.
+	 * 
+	 * @param context {@link ReviewContext}
+	 */
+	@After(entity = Books_.CDS_NAME)
+	public void afterAddReview(AddReviewContext context) {
+		String bookId = (String) analyzer.analyze(context.getCqn()).targetKeys().get(Books.ID);
+		bookRatingService.setBookRating(bookId);
 	}
 
 	@After(event = CdsService.EVENT_READ)
 	public void discountBooks(Stream<Books> books) {
-		books.filter(b -> b.getTitle() != null).forEach(
-			b -> {
-				loadStockIfNotSet(b);
-				discountBooksWithMoreThan111Stock(b);
+		books.filter(b -> b.getTitle() != null).forEach(b -> {
+			loadStockIfNotSet(b);
+			discountBooksWithMoreThan111Stock(b);
+		});
+	}
+
+	@After
+	public void setIsReviewable(CdsReadEventContext context, List<Books> books) {
+		String user = context.getUserInfo().getName();
+
+		Result result = db
```

Comment:
Could be improved if we only filter for the books that are acutally in the list, but I think for the demo case it is fine. Although maybe this could be a nice example for parameterized statements.